### PR TITLE
add generic paginator

### PR DESCRIPTION
**Pull request by zreigz**

### CODE REVIEW

File: algorithms/pager.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		HasNext:  true,
+		After:    nil,
+		PageSize: pageSize,
+	}
+	return p
+}
+
+func (p *Pager[T]) NextPage() ([]T, error) {
+	var list []T
+	for len(list) == 0 {
```
Comment (michaeljguarino) [inline algorithms/pager.go:36]: isn't this clause always true since you initialize list right before?

Comment (zreigz) [reply algorithms/pager.go]: oh, yes

Comment (zreigz) [reply algorithms/pager.go]: I've changed it

**Review by michaeljguarino — APPROVED** [review]

File: algorithms/pager.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+	if p.err != nil {
+		return nil, p.err
+	}
+
+	if !p.pageInfo.HasNext {
+		return nil, nil
+	}
+
+	list, p.pageInfo, p.err = p.fetchPage(p.pageInfo.After, p.pageInfo.PageSize)
```
Comment (michaeljguarino) [inline algorithms/pager.go:45]: Actually shouldn't we just return the error here directly as well instead of waiting until the next loop?

Comment (zreigz) [reply algorithms/pager.go:41]: yes, definitely

**Pull request merged**